File: Inversion/serial/NA/src/hv_subs/PROGRAMS.330/CALPLOT/testf/button.f
	subroutine mkmenu()
	common/menu/nm,mxl(5),myl(5),mxh(5),myh(5),mstr(5),mactio(5)
	integer*4 nm
	real*4 mxl, myl, mxh, myh
	character mstr*8
	integer*4 mactio
	nm = 5
	do 100 i=1,nm
		mxl(i) = -1
		myl(i) = -1
		mxh(i) = -1
		myh(i) = -1
		mactio(i) = i
  100	continue
	mstr(1) = '  Mode  '
	mstr(2) = ' Select '
	mstr(3) = ' Delete ' 
	mstr(4) = '  Copy  '
	mstr(5) = '  Exit  '
	return
	end


	program button
	common/ctrl/black
	integer*4 black
	common/menu/nm,mxl(5),myl(5),mxh(5),myh(5),mstr(5),mactio(5)
	integer*4 nm
	real*4 mxl, myl, mxh, myh
	character mstr*8
	integer*4 mactio

	real*4 xl,yl,xh,yh
	integer*4 i
	integer*4 cmd
	real*4 xv, yv
	character c*2
	character ostr*8
	call pinitf('INTER')
	call mkmenu()
	border = 8.0/60.0
	black = 1

	call drawit()
	
	do 1000 i=1,nm
		ostr = mstr(i)
		call draw_button(8.5,3.0 -i*0.4,ostr,xl,yl,xh,yh)
		mxl(i) = xl
		myl(i) = yl
		mxh(i) = xh
		myh(i) = yh
 1000	continue
c-----
c	event loop for values 
c-----
	cmd = -1
 2000	continue
		call gcursor('Arrow')
		call curaxy(xv, yv, c)
		cmd = -1
		do 2100 i=1, nm
			if(inside(xv,yv,mxl(i),myl(i),
     1				mxh(i),myh(i)).ne.0)then
     				cmd = mactio(i)
				call newpen(1)
				ostr = mstr(i)
				call gmesg(ostr)
			endif
 2100		continue
		if(cmd .eq. 5)then
			goto 9999
		else if(cmd .eq. 4)then
			call pinitf('button.plt')
			call drawit()
			call pend()
			call pinitf('INTER')
			call gcursor('Arrow')
		endif
	go to 2000
 9999	continue
	call gmesg('Closing Session - Any Key to End')
	call pend()
	end

	integer function inside( xv,  yv,  xlb, ylb,  xhb,  yhb)
	real*4 xv, yv, xlb, ylb, xhb, yhb
	if(xv .ge. xlb .and. xv .le. xhb .and. 
     1		yv .ge. ylb .and. yv .le. yhb)then
		inside = 1
	else
		inside = 0
	endif
	return
	end

	subroutine draw_button( xl,  yl, str,  xlw,  ylw, xup,  yup)
	real*4 xl, yl
	character str*(*)
	real*4 xlw, ylw, xup, yup
	common/ctrl/black
	integer*4 black
	integer*4 lstr
	integer*4 nc
	integer*4 IBcolr
	integer*4 IBLigh
	integer*4 IBDark
	integer*4 IBFore
	real*4 bdr, title, x(6), y(6)
	bdr = 0.04
	title = 20.0/60.0
	lstr = len(str)
c-----
c	add a space
c-----
	nc = lstr*8
	IBDark  = 1100
	IBcolr       = 1090
	IBLigh = 1070
c-----
c	color
c-----
	if(black.eq.1)then
		IBFore  =    0
	else
		IBFore  =    1
	endif
	call newpen(IBDark)
	x(1) = xl
	y(1) = yl
	x(2) = xl + lstr*0.125
	y(2) = yl
	x(3) = xl + lstr*0.125
	y(3) = yl + title 
	x(4) = xl + lstr*0.125 - bdr
	y(4) = yl + title - bdr 
	x(5) = xl + lstr*0.125 - bdr
	y(5) = yl + bdr 
	x(6) = xl + bdr
	y(6) = yl + bdr 
	call shadep(6,x,y)

	call newpen(IBLigh)
	x(1) = xl
	y(1) = yl
	x(2) = xl + bdr
	y(2) = yl + bdr 
	x(3) = xl + bdr
	y(3) = yl + title - bdr 
	x(4) = xl + lstr*0.125 - bdr
	y(4) = yl + title - bdr 
	x(5) = xl + lstr*0.125 
	y(5) = yl + title  
	x(6) = xl  
	y(6) = yl + title  
	call shadep(6,x,y)

	call newpen(IBcolr)
	x(1) = xl + bdr
	y(1) = yl + bdr
	x(2) = xl + lstr*0.125 - bdr
	y(2) = yl + bdr
	x(3) = xl + lstr*0.125 - bdr
	y(3) = yl + title - bdr
	x(4) = xl + bdr
	y(4) = yl + title - bdr
	call shadep(4,x,y)
	call newpen(IBFore)
	call gwrtxt(xl+0.0625,yl+0.015,str,0)
	xlw = xl
	ylw = yl
	xup = xl + lstr*0.125
	yup = yl + title
	return
	end

	subroutine drawit()
	integer*4 i, ipen
	real*4 xx, yy
	call box(3.0,4.0,6.0,6.0)
	call gclip('on',3.0,4.0,6.0,6.0)
	ipen = 3
	call newpen(4)
	do 1000 i=0, 100, 1
		xx = 3.0 + (i)*(6.0-3.0)/100.0
		yy = 5.0 + 2.0*sin(10.*xx)+cos(10.*xx)
		call plot(xx,yy,ipen)
		ipen = 2
 1000	continue
	call gclip('off',3.0,4.0,6.0,6.0)
	return
	end

	subroutine box(xl, yl, xh, yh)
	real*4 xl, yl, xh, yh
	call plot(xl,yl,3)
	call plot(xh,yl,2)
	call plot(xh,yh,2)
	call plot(xl,yh,2)
	call plot(xl,yl,2)
	call plot(xl,yl,3)
	return
	end

	function lgstr(str)
c-----
c	function to find the length of a string
c	this will only be used with file system path names
c	thus the first blank 
c	indicates the end of the string
c-----
	character*(*) str
	integer*4 lgstr
	n = len(str)
	lgstr = 1
	do 1000 i=n,1,-1
		lgstr = i
		if(str(i:i).ne.' ')go to 100
 1000	continue
  100   continue
        return
	end

File: Inversion/parallel/NA/src/hv_subs/PROGRAMS.330/CALPLOT/testc/button.c
#include	<stdio.h>
#include "calplot.h"
#include <math.h>

float border;
float title;
int Button_Color;
int Button_Color_Light;
int Button_Color_Dark;
int Button_Color_Fore;
int kolor;
int black;
void box(float xl, float yl, float xh, float yh);
void draw_button(float xl, float yl, char *str, float *xlw, float *ylw,
		float *xup, float *yup);
static int inside(float xv, float yv, 
	float xlb, float ylb, float xhb, float yhb);
struct menu { float xl; float yl; float xh 
	; float yh ; char *str ; int action;  } ;
static struct menu m[] = { 
	{  -1.0, -1.0, -1.0, -1.0, "  Mode  \0" , 1},
	{  -1.0, -1.0, -1.0, -1.0, " Select \0" , 2},
	{  -1.0, -1.0, -1.0, -1.0, " Delete \0" , 3},
	{  -1.0, -1.0, -1.0, -1.0, "  Copy  \0" , 4},
	{  -1.0, -1.0, -1.0, -1.0, "  Exit  \0" , 5}
} ;

FILE *buterr;

 main()
{
	float xl,yl,xh,yh;
	int nm, i;
	int cmd;
	float xv, yv;
	char c[2];
	buterr= fopen("button.err","w");
	ginitf("INTER","BUTTON");
	border = 8.0/60.0;
	kolor = 1;
	black = 1;

	drawit();
	nm = sizeof(m)/sizeof(struct menu);
	for(i=0 ; i < nm ; i++){
		draw_button(8.5,3.0 -i*0.4,m[i].str,&xl,&yl,&xh,&yh);
		m[i].xl = xl;
		m[i].yl = yl;
		m[i].xh = xh;
		m[i].yh = yh;
	}
	/* loop for values */
	cmd = -1;
	for(; ;){
		gcursor("Arrow");
		curaxy(&xv, &yv, c);
		cmd = -1;
		for(i=0 ; i < nm ; i++){
			if(inside(xv,yv,m[i].xl,m[i].yl,m[i].xh,m[i].yh))
			{
				cmd = m[i].action;
				gmesg(m[i].str);
			}
		}
		if(cmd == 5)
			goto end;
		else if(cmd == 4){
			pinitf("BUTTON.PLT");
			drawit();
			pend();
			ginitf("INTER","BUTTON");
			gcursor("Arrow");
		}
	}
end:
	gmesg("Closing Session - Any Key to End");
	pend();
fclose(buterr);
}

static int inside(float xv, float yv, float xlb, 
	float ylb, float xhb, float yhb)
{
	if(xv >= xlb && xv <= xhb && yv >= ylb && yv <= yhb)
		return 1;
	else
		return 0;
}

void draw_button(float xl, float yl, char *str, float *xlw, float *ylw,
		float *xup, float *yup)
{
	int lstr;
	int nc;
	float bdr = 0.04;
	float title = 20.0/60.0;
	float x[6], y[6];
	lstr = strlen(str);
	/* add a space */
	nc = lstr*8;
		Button_Color_Dark  = 1100;
		Button_Color       = 1090;
		Button_Color_Light = 1070;
		/* color */
		if(black){
			Button_Color_Fore  =    0;
		} else {
			Button_Color_Fore  =    1;
		}
	newpen(Button_Color_Dark);
	x[0] = xl;
	y[0] = yl;
	x[1] = xl + lstr*0.125;
	y[1] = yl;
	x[2] = xl + lstr*0.125;
	y[2] = yl + title ;
	x[3] = xl + lstr*0.125 - bdr;
	y[3] = yl + title - bdr ;
	x[4] = xl + lstr*0.125 - bdr;
	y[4] = yl + bdr ;
	x[5] = xl + bdr;
	y[5] = yl + bdr ;
	shadep(6,x,y);

	newpen(Button_Color_Light);
	x[0] = xl;
	y[0] = yl;
	x[1] = xl + bdr;
	y[1] = yl + bdr ;
	x[2] = xl + bdr;
	y[2] = yl + title - bdr ;
	x[3] = xl + lstr*0.125 - bdr;
	y[3] = yl + title - bdr ;
	x[4] = xl + lstr*0.125 ;
	y[4] = yl + title  ;
	x[5] = xl  ;
	y[5] = yl + title  ;
	shadep(6,x,y);

	newpen(Button_Color);
	x[0] = xl + bdr;
	y[0] = yl + bdr;
	x[1] = xl + lstr*0.125 - bdr;
	y[1] = yl + bdr;
	x[2] = xl + lstr*0.125 - bdr;
	y[2] = yl + title - bdr;
	x[3] = xl + bdr;
	y[3] = yl + title - bdr;
	shadep(4,x,y);
	newpen(Button_Color_Fore);
	gwrtxt(xl+0.0625,yl+0.015,str,0);
	*xlw = xl;
	*ylw = yl;
	*xup = xl + lstr*0.125;
	*yup = yl + title;
}

drawit()
{
	int ix, iy;
	char ch[2];
	int i, ipen;
	float xx, yy;
	box(3.0,4.0,6.0,6.0);
	gclip("on",3.0,4.0,6.0,6.0);
	ipen = 3;
	newpen(4);
	for(i=0;i<= 100 ; i++){
		xx = 3.0 + (i)*(6.0-3.0)/100.0;
		yy = 5.0 + 2.0*sin(10.*xx)+cos(10.*xx);
		plot(xx,yy,ipen);
		ipen = 2;
	}
	gclip("off",3.0,4.0,6.0,6.0);
}

void box(float xl, float yl, float xh, float yh)
{
	plot(xl,yl,3);
	plot(xh,yl,2);
	plot(xh,yh,2);
	plot(xl,yh,2);
	plot(xl,yl,2);
	plot(xl,yl,3);
}
